### PR TITLE
Reconnect to remote devbox agent when connection is dropped

### PR DESCRIPTION
**Pull request by codeyq**
This PR adds networking check and retry logic in devbox syncer. When syncer haven't received any ack from remote agent, it will retry on the networking connection and flush pending buffers.

### CODE REVIEW

File: devbox/common/src/RpcClient.scala [hunk 1]
```diff
@@ -1,20 +1,44 @@
...
-                in: InputStream with DataInput,
-                logger: (String, Any) => Unit) {
+import scala.util.control.NonFatal
+
+class RpcClient(var out: OutputStream with DataOutput,
+                var in: InputStream with DataInput,
+                logger: (String, Any) => Unit,
+                ackPing: Option[() => Unit] = None) {
   private[this] val outstandingMsgs = new AtomicInteger()
+  private[this] val pendingQueue = new ConcurrentHashMap[Int, Rpc]()
```
Comment (lihaoyi-databricks) [inline devbox/common/src/RpcClient.scala]: `java.util.ConcurrentLinkedHashMap`

**Comment by lihaoyi-databricks** [comment]
As we discussed over hangouts, here are the main pieces of feedback:

- Let's move `flushOutstandingMsgs` from the `ScheduledThreadPoolExecutor` into the start of the main thread's `while(continue())` loop. This will allow us to enforce that flushing existing messages and sending new messages never happen concurrently, ruling out a whole class of potential bugs

- `pendingQueue` can be a `Queue` (or `BlockingQueue` or `ConcurrentLinkedQueue` or something) rather than a `ConcurrentHashMap`. This will allow us to preserve ordering, such that when we replay the messages. Since both the main `Syncer` logic and remote `DevboxAgentMain` are both single threaded, and exchanging messages of TCP (which enforces in-order delivery) we can be confident that any `Ack`s that get received will be for the entries at the head of the queue.

- We should preserve the behavior of the `draining(client){...}` blocks, since that is what allows us to ensure things have fully synced before showing notifications to the user. We can do this by waiting until the `pendingQueue` has emptied out before returning, OR the agent has died in which case you bail out early and make another go around the big `while(continue())` loop

- This was not discussed, but I suspect we should also move the logic for re-creating the agent (`agent.start()`, client.resetInt`, `client.resetOut`) to the start of the big `while(continue())` loop. This will ensure that if the agent dies halfway through the main thread syncing something, it stays dead until the main thread comes around again, reducing the potential for bugs in the following case: the agent dies halfway through the big loop, the main thread sends some messages that go into `pendingQueue`, the agent restarts, and the main thread continues sending new messages while there are still old messages sitting in `pendingQueue`, meaning the messages will arrive out of order.

- As discussed, we should have a standalone automated test for this recovery behavior. You can run the existing tests via `mill devbox.test devbox.DevboxTests.edge.0`, `mill devbox.test devbox.DevboxTests.edge.restart`, `mill devbox.test devbox.DevboxTests.scalatags.0`, etc.. We already have handy mock Agents for testing, e.g. `InMemoryAgent` or a local subprocess syncing between two folders (see `def instantiateSyncer`), and it shouldn't be too hard to inject some failures at the `Agent`-level that we expect `Syncer.scala` to be able to recover from.

These are the high level pieces of feedback; nitpicks can come later. Thanks for doing this!

**Comment by codeyq** [comment]
Thanks @lihaoyi-databricks 
- This is hard because the drainning thread is not always running (only runs in `synchronizeRepo`) Therefore, if connection died when draining thread is running, main thread should be stuck at `drainer.join` and won't be able to execute `flushOutstandingMsgs`. My current solution is that when `monitor` detects that the connection is dropped and attemp to reconnect, it set a state in client `shouldFlush`. In the draining thread, it checks if `shouldFlush` and flushes outstanding messages. This is safe because draining thread will only finish when all outstanding messages are acked.
- Done
- Done. One thing should be careful is that `Pong` is received by reading message from remote agent which is performed by a draining thread. I added another `DevboxBackgroundDrainerThread` which is only running when no other `DevboxDrainerThread` is running.
- This is no longer the case because we ensure that `draining` is a checkpoint. Therefore, if agent dies halfway and restarts, main thread will flush all outstanding messages first, join the draining thread and send new messages.

**Comment by lihaoyi-databricks** [comment]
> This is hard because the drainning thread is not always running (only runs in synchronizeRepo) Therefore, if connection died when draining thread is running, main thread should be stuck at drainer.join and won't be able to execute flushOutstandingMsgs. My current solution is that when monitor detects that the connection is dropped and attemp to reconnect, it set a state in client shouldFlush. In the draining thread, it checks if shouldFlush and flushes outstanding messages. This is safe because draining thread will only finish when all outstanding messages are acked.

This seems reasonable to me: so if the connection dies, the main thread's messages will get ignored by the dead subprocess, and the thread will get "stuck" at the conclusion of `draining{...}`. The main thread will thus stay there until the connection is re-established, at which point the drainer thread will notice it `shouldFlush`, and will `flushOutstandingMsgs` before terminating and letting the main thread proceed.

Do I understand that correctly?

**Comment by codeyq** [comment]
> > This is hard because the drainning thread is not always running (only runs in synchronizeRepo) Therefore, if connection died when draining thread is running, main thread should be stuck at drainer.join and won't be able to execute flushOutstandingMsgs. My current solution is that when monitor detects that the connection is dropped and attemp to reconnect, it set a state in client shouldFlush. In the draining thread, it checks if shouldFlush and flushes outstanding messages. This is safe because draining thread will only finish when all outstanding messages are acked.
> 
> This seems reasonable to me: so if the connection dies, the main thread's messages will get ignored by the dead subprocess, and the thread will get "stuck" at the conclusion of `draining{...}`. The main thread will thus stay there until the connection is re-established, at which point the drainer thread will notice it `shouldFlush`, and will `flushOutstandingMsgs` before terminating and letting the main thread proceed.
> 
> Do I understand that correctly?

Exactly. This is a bit messy to spread state all around different component and makes it very hard to maintain (I will put more comments in the PR). But I found it the simplest way to keep `draining` semantics while not refactoring a huge chunk of code.

**Comment by lihaoyi-databricks** [comment]
@codeyq I put some time on your calendar to go over this PR together, but we could also do it now if you're free/online

**Comment by haogang** [comment]
Shall we consider replace ssh with mosh (https://mosh.org/)? That should solve the auto reconnect problem when the client switches to another network. It also has other perks like local echo to improve interactivity when the network latency is high.

**Comment by codeyq** [comment]
@haogang I think this PR tends to make sure the underlying connection is stable no matter what kind of connection because devbox only cares about stdout, stdin and stderr interface with the remote agent.
Currently, we run `devbox.jar` which takes a ssh command as a parameter. If we want to use mosh, we can just switch this ssh one to mosh.
```bash
java -jar devbox.jar \
     --log-file ../log.txt \
     --healthcheck-interval 3 \
     --repo /Users/qianyu/devbox ssh -p 2222 root@localhost java -jar /root/agent.jar
```

**Comment by lihaoyi-databricks** [comment]
@haogang this is for the syncer thread, not the SSH thread, so MOSH's local echo wouldn't be useful. If you want to use MOSH on the devbox interactively, you can install it and connect manually (or add it to our `setup_devbox.sh` script so it's always installed by default). 

I'm not sure how well MOSH does with Devbox's command connection, since it's sending over large blobs of bytes rather than the interactive terminal sessions that MOSH is designed to optimize for, but its' worth trying as qian mentioned above

**Comment by lihaoyi-databricks** [comment]
This generally looks good to me. Last comments:

- Let's go make the `DevboxAgentMain` RPC handlers idempotent. We should make it idempotent when deployed, but turn off the idempotency in devbox's own test suite so the tests can be more aggressive at finding bugs

- Clean up the various `.sh` scripts you were using for testing

- Maybe revert https://github.com/databricks/devbox/pull/1/files#diff-b77033c1e4cce42c184d0d0cc278cdd9L11 if possible

**Pull request merged**